packages/rfw/test/runtime_test.dart

```diff
@@ -1088,4 +1088,156 @@ void main() {
+
+      final Finder textFinder = find.byType(Text);
+      expect(textFinder, findsOneWidget);
+      expect(tester.widget<Text>(textFinder).data, 'Hello Widget Builders!');
+    });
+
+    testWidgets('Widget builders - work when scope is used', (WidgetTester tester) async {
+      final Widget widget = _setUp('''
+        import core;
+        import local;
+
+        widget test = Calculator(
+          operand1: 1,
+          operand2: 2,
+          operation: 'sum',
+          builder: (result) => CoolText(text: ['1 + 2 = ', result.result]),
+        );
+      ''');
+      await tester.pumpWidget(widget);
+
+      final Finder textFinder = find.byType(Text);
+      expect(textFinder, findsOneWidget);
+      expect(tester.widget<Text>(textFinder).data, '1 + 2 = 3');
+    });
+
+    testWidgets('Widget builders - works nested', (WidgetTester tester) async {
+      final Widget widget = _setUp('''
+        import core;
+        import local;
+
+        widget test = Calculator(
+          operand1: 1,
+          operand2: 2,
+          operation: 'sum',
+          builder: (result1) => Calculator(
+            operand1: result1.result,
+            operand2: 3,
+            operation: 'multiply',
+            builder: (result2) => CoolText(text: ['(1 + 2) * 3 = ', result2.result]),
+          ),
+        );
+      ''');
+      await tester.pumpWidget(widget);
+
+      final Finder textFinder = find.byType(Text);
+      expect(textFinder, findsOneWidget);
+      expect(tester.widget<Text>(textFinder).data, '(1 + 2) * 3 = 9');
+    });
+
+    testWidgets('Widget builders - swich works with builder', (WidgetTester tester) async {
+      final Widget widget = _setUp('''
+        import core;
+        import local;
+
+        widget test {justResult: true} = Calculator(
+          operand1: 1,
+          operand2: 2,
+          operation: 'sum',
+          builder: switch state.justResult {
+            true: (result) => CoolText(text: [result.result]),
+          },
+        );
+      ''');
+      await tester.pumpWidget(widget);
+
+      final Finder textFinder = find.byType(Text);
+      expect(textFinder, findsOneWidget);
+      expect(tester.widget<Text>(textFinder).data, '3');
+    });
+
+    testWidgets('Widget builders - builder works with switch', (WidgetTester tester) async {
+      final Widget widget = _setUp('''
+        import core;
+        import local;
+
+        widget test = Calculator(
+          operand1: 1,
+          operand2: 2,
+          operation: 'sum',
+          builder: (result) => switch result.result {
+            0: CoolText(text: ['The result is zero']),
+            default: CoolText(text: ['The result is not zero']),
+          },
+        );
+      ''');
+      await tester.pumpWidget(widget);
+
+      final Finder textFinder = find.byType(Text);
+      expect(textFinder, findsOneWidget);
+      expect(tester.widget<Text>(textFinder).data, 'The result is not zero');
+    });
+  });
 }
```

Comment:
these tests are great. we should also add some tests that verify that things update correctly when the data is updated. for example, have a version of the calculator test where the operands come from a data reference, and the data is updated several times and you see if that causes the render tree to change appropriately.

Reply:
Great call @Hixie, done!